src/app/interaction-model/InvokeResponder.h

```diff
@@ -0,0 +1,192 @@
+    /// If this returns CHIP_ERROR_BUFFER_TOO_SMALL, this can be called a 2nd time after
+    /// a FlushPendingReplies.
+    ///
+    /// Argument behavior:
+    ///  - If an ERROR is given (i.e. NOT CHIP_NO_ERROR) an error response will be given to the
+    ///    command.
+    ///  - If CHIP_NO_ERROR is given:
+    ///    - if a ReplyEncoder() was called, a data reply will be sent
+    ///    - if no ReplyEncoder() was called, a Success reply will be sent
+    ///
+    /// Returns success/failure state. One error code MUST be handled in particular:
+    ///
+    ///   - CHIP_ERROR_BUFFER_TOO_SMALL will return IF AND ONLY IF the responder was unable
+    ///     to fully serialize the given reply/error data.
+    ///
+    ///     If such an error is returned, the caller MUST retry by calling
+    ///
+    virtual CHIP_ERROR Complete(CHIP_ERROR error) = 0;
+};
+
+/// Enforces that once acquired, Complete will be called on the underlying writer
+class AutoCompleteInvokeResponder
+{
+public:
+    // non-copyable: once you have a handle, keep it
+    AutoCompleteInvokeResponder(const AutoCompleteInvokeResponder &)             = delete;
+    AutoCompleteInvokeResponder & operator=(const AutoCompleteInvokeResponder &) = delete;
+
+    AutoCompleteInvokeResponder(InvokeResponder * writer) : mWriter(writer) {}
+    ~AutoCompleteInvokeResponder()
+    {
+        if (!mCompleted)
+        {
+            mWriter->Complete(CHIP_IM_GLOBAL_STATUS(Failure));
+        }
+    }
+
+    /// Direct access to reply encoding.
+    ///
+    /// Use this only in conjunction with the other Raw* calls
+    DataModel::WrappedStructEncoder & RawReplyEncoder(CommandId replyCommandId) { return mWriter->ReplyEncoder(replyCommandId); }
+
+    /// Direct access to flushing replies
+    ///
+    /// Use this only in conjunction with the other Raw* calls
+    CHIP_ERROR RawFlushPendingReplies() { return mWriter->FlushPendingReplies(); }
+
+    /// Call "Complete" without the automatic retries.
+    ///
+    /// Use this in conjunction with the other Raw* calls
+    CHIP_ERROR RawComplete(CHIP_ERROR error) { return mWriter->Complete(error); }
+
+    /// Complete the given command.
+    ///
+    /// Automatically handles retries for sending.
+    CHIP_ERROR Complete(CHIP_ERROR error)
+    {
+        mCompleted     = true;
+        CHIP_ERROR err = mWriter->Complete(error);
+
+        if (err != CHIP_ERROR_BUFFER_TOO_SMALL)
+        {
+            return err;
+        }
+
+        // retry once. Failure to flush is permanent.
+        ReturnErrorOnFailure(mWriter->FlushPendingReplies());
+        return mWriter->Complete(error);
+
+        return err;
+    }
+
+    /// Sends the specified data structure as a response
+    ///
+    /// This version of the send has built-in RETRY and handles
+    /// Flush/Complete automatically.
+    ///
+    template <typename ReplyData>
+    CHIP_ERROR Send(const ReplyData & data)
+    {
```

Comment:
Should you check if `mCompleted` is already set to true here?

Also should you be setting it to true somewhere in the method?

Reply:
Added checks and additional sets/resets.